File: src/main/java/org/fountainmc/api/entity/data/hanging/ItemFrameData.java
package org.fountainmc.api.entity.data.hanging;

import javax.annotation.Nullable;

import org.fountainmc.api.Material;
import org.fountainmc.api.entity.data.EntityData;

public interface ItemFrameData extends HangingEntityData {

    /**
     * Does the ItemFrame contain an Item?
     *
     * @return whether the ItemFrame contains an Item
     */
    default boolean containsItem() {
        return getMaterial() != null;
    }

    /**
     * Get the Material of the Item in the ItemFrame, or null if there isn't an item.
     *
     * @return the Material of the Item in the ItemFrame
     */
    @Nullable
    Material getMaterial();

    /**
     * Set the Material of the Item in the ItemFrame.
     * <p>Setting it to null empties the item frame.</p>
     *
     * @param material the Material to set the Item in the ItemFrame to
     */
    void setMaterial(@Nullable Material material);

    @Override
    default void copyDataFrom(EntityData data) {
        HangingEntityData.super.copyDataFrom(data);
        if (data instanceof ItemFrameData) {
            this.setMaterial(((ItemFrameData) data).getMaterial());
        }
    }
}

File: src/main/kotlin/org/fountainmc/api/entity/data/hanging/ItemFrameData.kt
package org.fountainmc.api.entity.data.hanging

import org.fountainmc.api.Material
import org.fountainmc.api.entity.data.EntityData

interface ItemFrameData : HangingEntityData {

    /**
     * If the ItemFrame contains an Item
     */
    val containsItem
        get() = material != null

    /**
     * The Material of the Item in the ItemFrame, or null if there isn't an item.
     */
    var material: Material<*>?

    override fun copyDataFrom(data: EntityData) {
        super.copyDataFrom(data)
        if (data is ItemFrameData) {
            this.material = data.material
        }
    }
}
